### PR TITLE
Allow `#[must_use]` on functions, rather than just types. Mark `Result::{ok,err}` `#[must_use]`.

### PR DESCRIPTION
**Pull request by huonw**
[Rendered](https://github.com/huonw/rfcs/blob/must-use/text/0000-must-use-functions.md).

### CODE REVIEW

File: text/0000-must-use-functions.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    Bar.baz(); // warning: unused result that must be used: maybe you meant something else
+}
+```
+
+
+# Drawbacks
+
+This adds a little more complexity to the `#[must_use]` system.
+
+The rule stated doesn't cover every instance were a `#[must_use]`
```
Comment (Gankro) [inline text/0000-must-use-functions.md]: s/were/where

File: text/0000-must-use-functions.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+function is ignored (it does cover all instances of a `#[must_use]`
+type), e.g. `(foo());` and `{ ...; foo() };` will not be picked up
+(that is, passing the result through another piece of syntax). This
+could be tweaked.
```
Comment (Gankro) [inline text/0000-must-use-functions.md]: This paragraph is kinda hard to parse (two asides in the same "sentence")

Comment (huonw) [inline text/0000-must-use-functions.md]: Adjusted.

File: text/0000-must-use-functions.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+`Result::ok` is occasionally used for silencing the `#[must_use]`
+error of `Result`, i.e. the ignoring of `foo().ok();` is
+intentional. However, the most common way do ignore such things is
+with `let _ =`, and `ok()` is rarely used in comparison, in most
+code-bases: 2 instances in the rust-lang/rust codebase (vs. nearly 400
+text matches for `let _ =`) and 4 in the servo/servo (vs. 55 `let _
+=`). Yet another way to write this is `drop(foo())`, although neither
+this nor `let _ =` have the method chaining style.
```
Comment (Gankro) [inline text/0000-must-use-functions.md]: Is it within scope for this RFC to try to settle on a convention?

File: text/0000-must-use-functions.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+# Alternatives
+
+- Adjust the rule to propagate `#[must_used]`ness through parentheses
+  and blocks, so that `(foo());`, `{ foo() };` and even `if cond {
+  foo() } else { 0 };` are linted.
+
+- Provide an additional method on `Result`, e.g. `fn ignore(self) {}`, so
+  that users who wish to ignore `Result`s can do so in the method
+  chaining style: `foo().ignore();`.
```
Comment (Gankro) [inline text/0000-must-use-functions.md]: :+1: from me. Hooray for semantics (and easy to grep/lint for once you're ready for production!)

File: text/0000-must-use-functions.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+code-bases: 2 instances in the rust-lang/rust codebase (vs. nearly 400
+text matches for `let _ =`) and 4 in the servo/servo (vs. 55 `let _
+=`). Yet another way to write this is `drop(foo())`, although neither
+this nor `let _ =` have the method chaining style.
+
+# Alternatives
+
+- Adjust the rule to propagate `#[must_used]`ness through parentheses
+  and blocks, so that `(foo());`, `{ foo() };` and even `if cond {
+  foo() } else { 0 };` are linted.
```
Comment (Gankro) [inline text/0000-must-use-functions.md]: Any idea how complicated this would be to do?

File: text/0000-must-use-functions.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+- Provide an additional method on `Result`, e.g. `fn ignore(self) {}`, so
+  that users who wish to ignore `Result`s can do so in the method
+  chaining style: `foo().ignore();`.
+
+# Unresolved questions
+
+- Are there many other functions in the standard library/compiler
+  would benefit from `#[must_use]`?
```
Comment (Gankro) [inline text/0000-must-use-functions.md]: Its been suggested elsewhere that the comparison operators should be linted thusly. No reasonable code should have `x == y;` as a standalone expression.

Would it be possible to mark a *trait's* method as must_use, affecting all implementers?

File: text/0000-must-use-functions.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+The `#[must_use]` lint is extremely useful for ensuring that values
+that are likely to be important are handled, even if by just
+explicitly ignoring them with, e.g., `let _ = ...;`. This expresses
+the programmers intention clearly, so that there is less confusion
+about whether, for example, ignoring the possible error from a `write`
+call is intentional or just an accidental oversight.
+
+Rust has got a lot of mileage out connecting the `#[must_use]` lint to
+specific types: types like `Result`, `MutexGuard` (any guard, ina
```
Comment (reem) [inline text/0000-must-use-functions.md]: typo at ina

File: text/0000-must-use-functions.md (hunk 5 above)
Comment (eddyb) [inline text/0000-must-use-functions.md]: I think the easiest thing to do would be to track whether the result of evaluating the current expression is obviously unused - that would propagate down through `(foo())`, `{ foo() }`, `if`/`else`, `match`, etc.

File: text/0000-must-use-functions.md (hunk 4 above)
Comment (reem) [inline text/0000-must-use-functions.md]: Agreed. I like this solution, though am also in favor of the RFC as a generally useful construct.

Comment (Gankro) [inline text/0000-must-use-functions.md]: Although `ignore` as a method is a bit wanting if any return type can be marked as must_use. Still need a general convention (unless Result is the only one that "needs" ignore-ability).

**Comment by mahkoh** [comment]
-1. I, and most code-bases, use ok() to ignore Result.

**Comment by huonw** [comment]
As stated in the RFC, there's not a huge amount of evidence for "most".

**Comment by mahkoh** [comment]
You know what they say: Absence of evidence is not evidence of absence.

**Comment by eddyb** [comment]
:+1: Seeing `.ok()` *anywhere* is still quite confusing for me.
`.ok().unwrap()` makes me wonder what is wrong with their error type that results in `.unwrap()` not being used.
And `foo().ok();` was "obviously" meant to be used and the author *must* have forgotten to do something with it.
Very rarely have I seen an use I consider legitinately, and even then a match or if-let would be clearer.
@nikomatsakis I know you've been using `Result` in rustc algorithms a lot lately, what are your thoughts on the matter?

**Comment by huonw** [comment]
@mahkoh I have looked, I have found evidence of absence. If you can find a wealth of libraries from a variety of authors that use `.ok()` for ignoring `Result`s the position of the RFC can be reconsidered, but, at the moment, using `.ok();` for this purpose seems to be quite unidiomatic (or, at least, uncommon).

Even then, I would prefer having an explicit `.ignore()` method than abusing `.ok()` for this purpose, since the former clearly states what it is for while the latter has the problems described in the motivation section.

**Comment by mahkoh** [comment]
> I have looked, I have found evidence of absence.

That's great! I'm looking forward to your analysis of most code-bases on crates.io. Btw: Looking at two trees and not finding any squirrels does not extrapolate to "there are no squirrels in the forrest."

> If you can find

Let's not try to move the burden of proof. You're the one proposing a change and making claims about "ok" not being used for this purpose in most code-bases.

>Even then, I would prefer having an explicit `.ignore()`

Then we [agree](https://github.com/rust-lang/rust/issues/20949). Unfortunately there doesn't seem to be a consensus for changing `ok` or adding `ignore`.

**Comment by huonw** [comment]
```
$ ack-grep --type rust 'let _ =' ~/.cargo ~/.multirust/toolchains/*/cargo | wc -l
533
$ ack-grep --type rust '\.ok\(\);' ~/.cargo ~/.multirust/toolchains/*/cargo | grep -v '=.*ok' | wc -l
64
```

**Comment by mahkoh** [comment]
Is that your detailed analysis of most code-bases on crates.io?

**Comment by huonw** [comment]
Clearly not, but the claim "most code-bases use ok() to ignore Result." is looking pretty unlikely.

**Comment by mahkoh** [comment]
```
~/rust$ grep "let _ =" $(find -name "*.rs") | wc -l 
17
~/rust$ grep "\.ok()" $(find -name "*.rs") | wc -l 
30
```

Please focus on proving your claims for now. In case you have forgotten:

>`Result::ok` is occasionally used for silencing the `#[must_use]`
>error of `Result`, i.e. the ignoring of `foo().ok();` is
>intentional. However, the most common way do ignore such things is
>with `let _ =`, and `ok()` is rarely used in comparison, in most
>code-bases

File: text/0000-must-use-functions.md (hunk 5 above)
Comment (oli-obk) [inline text/0000-must-use-functions.md]: So basically automatic derivation of must_use for all types (if you have a sub-element that has must_use, you are must_use)? any expression doing a no-op on the object will also return a must_use object -> error message is forwarded outside the expression.

**Comment by mahkoh** [comment]
If the first part of this RFC (#[must_use] on functions) is implemented, then #[must_use] should be removed from Result.

**Comment by oli-obk** [comment]
I think it's irrelevant whether `ok()` is used often or not. It has a purpose: converting a `Result` into an `Option`. `let _ =` has no other purpose than ignoring results. A rule of thumb in safety is to keep dangerous and safe actions as far away from each other as possible. Exaggerated: "Don't put the nuclear bomb launch button next to the coffee button".

**Comment by oli-obk** [comment]
>If the first part of this RFC (#[must_use] on functions) is implemented, then #[must_use] should be removed from Result.

That would require every function that returns a Result to be annotated. Some might be missed. The Result type exists explicitly for two things: error reporting and ensuring errors don't get lost.

**Comment by mahkoh** [comment]
>That would require every function that returns a Result to be annotated

That's obviously not true. It only requires functions where the return value must be used to be annotated. E.g. if a function is changed from
```rust
fn f() -> Option<T>;
```
to
```rust
fn f() -> Result<T, ()>;
```
then that does not imply that ignoring the return value suddenly becomes more or less dangerous.

**Comment by huonw** [comment]
> `~/rust$ grep "let _ =" $(find -name "*.rs") | wc -l `

AFAIK, that is the code of a single author who is clearly *very* attached to using `ok()` for this purpose (why else would you be fighting so hard?). Further, `ok()` has actual non-ignoring uses (when someone wants and uses the returned `Option`), which you `grep` does not pick up. I've been careful to use `.ok();` (with the semicolon) to try to focus on the interesting cases.

If "most" code bases use it for this purpose, it shouldn't be too hard for you to find a few on github to point me to.

I did my own random sample of 50 projects on Rust CI, and added a write up with statistical analysis, to the end of the RFC. I'm now convinced that this will have minimal impact on most code. Unless I see some very strong evidence to the contrary, I'm considering the objection that `ok` is used a lot for this purpose void.

> If the first part of this RFC (#[must_use] on functions) is implemented, then #[must_use] should be removed from Result.

I'm slow, and your customarily terse comments are not at all self-explanatory. Please expand.

**Comment by huonw** [comment]
The data table looks best [rendered](https://github.com/huonw/rfcs/blob/must-use/text/0000-must-use-functions.md#data).

**Comment by oli-obk** [comment]
> E.g. if a function is changed
Then you are breaking the public api of your library.

> then that does not imply that ignoring the return value suddenly becomes more or less dangerous.
Yes... Isn't that why you went to use `Result` instead of `Option`? Error reporting and handling?

**Comment by mahkoh** [comment]
>AFAIK, that is the code of a single author

You are wrong.

>it shouldn't be too hard for you to find a few on github to point me to.

That would imply that I agree to shift the burden of proof. The examples above was merely a quick way to show that your "grep-analysis" was flawed.

>I'm slow, and your customarily terse comments are not at all self-explanatory. Please expand.

```
diff --git a/src/libcore/result.rs b/src/libcore/result.rs
index d610962..a96a1fa 100644
--- a/src/libcore/result.rs
+++ b/src/libcore/result.rs
@@ -240,7 +240,6 @@ use slice;
 ///
 /// See the [`std::result`](index.html) module documentation for details.
 #[derive(Clone, Copy, PartialEq, PartialOrd, Eq, Ord, Debug, Hash)]
-#[must_use]
 #[stable(feature = "rust1", since = "1.0.0")]
 pub enum Result<T, E> {
     /// Contains the success value
```

>I did my own random sample of 50 projects on Rust CI, and added a write up, with statistical analysis, to the end of the RFC. I'm now convinced that this will have minimal impact on most code. Unless I see some very strong evidence to the contrary, I'm considering the objection that ok is commonly used for this purpose just outright wrong.

I have not verified your results but from the data alone I have to agree with you!

**Comment by mahkoh** [comment]
>Then you are breaking the public api of your library.

Not all functions in a library are public and not all functions are in libraries.

>Yes... Isn't that why you went to use Result instead of Option? Error reporting and handling?

Not at all! I use Result<T, ()> quite often in place of Option<T> because it comes with the handy try! macro.

**Comment by oli-obk** [comment]
> Not at all! I use Result<T, ()> quite often in place of Option<T> because it comes with the handy try! macro.

While I see the need for a new macro for options... This is abuse... I surrender

**Comment by huonw** [comment]
> You are wrong.

I would appreciated it if you could have provided enough information to know this ahead of time.

> That would imply that I agree to shift the burden of proof. The examples above was merely a quick way to show that your "grep-analysis" was flawed.

The analysis I had done previously (a 10x ratio between `ok` and `let _ =` for the crates in my cargo caches) was easily more than enough for there to be some "burden of proof" on you.

Just because I'm proposing changes doesn't mean that objections don't need proof.

> `-#[must_use]`

That is clearly not what I was asking about. Why "should"?

---

@mahkoh I'm getting *very* tired of having to repeatedly ask you to expand comments and provide more context, in RFCs, in filed issues, and even when you ask for people to assist you on IRC; it's like pulling teeth. Furthermore, your objections (to RFCs and people's comments) are often so brief as to be borderline [unstructured cricitism](http://www.rust-lang.org/conduct.html).

**Comment by oli-obk** [comment]
@huonw: could you address must_use traits (or at least forbid them explicitly for now)? Currently that attribute on traits does nothing (see http://is.gd/JMciVr)

**Comment by mahkoh** [comment]
>I would appreciated it if you could have provided enough information to know this ahead of time.

You did not bother to show us the contents of your .cargo folders when you presented your quick analysis.

>Why "should"?

Because there is a clear bias towards using Result for return values that can contain success and error values:

- It's in the prelude
- It's in the stdlib and therefore better for interop
- It occupies the try! macro

For these reasons people might choose to use the Result type instead of writing their own Either\<A, B\> even though the rv is not must_use.

In general: It seems that must_use was mostly motivated by the use of Result in IO functions. With this RFC this becomes redundant because IO functions themselves can be tagged must_use.

This seems pretty clear.

**Comment by nikomatsakis** [comment]
Wow, I am surprised @mahkoh to see you so vociferously defending the idea of writing `result.ok()` as a way of indicating that a Result is not to be inspected. It seems to me to be self-evidently non-optimal:

1. It is not obvious what is going on. It reads like you're asserting that the result is ok or something.
2. It is inefficient: you are testing and creating an Option. Maybe (probably) LLVM will optimize it away, but why force LLVM to do that? And I wouldn't be sure of that in all cases.

Based on these reasons alone, I personally think that if there is a lot of code "in the wild" using this pattern, it ought to be updated in any case. Unsurprisingly, @huonw's data seems to show that there is not, so it doesn't matter.

**Comment by nikomatsakis** [comment]
In general I am in favor of this proposal. It seems to make `#[must_use]` strictly more useful. I'd like to see some effort put into handling things like `(x.foo())`, though it'll no doubt never be perfect, but that's ok.

**Comment by mahkoh** [comment]
>It is not obvious what is going on. It reads like you're asserting that the result is ok or something.

Which is why is suggested an `ignore` function some time ago. Same functionality, same ergonomics, more obvious name. Unfortunately this RFC doesn't do that.

>It is inefficient: you are testing and creating an Option. Maybe (probably) LLVM will optimize it away, but why force LLVM to do that? And I wouldn't be sure of that in all cases.

There are much bigger fish to fry than that.

**Comment by shepmaster** [comment]
This proposal seems quite useful, and I am :+1: . Some of the more-awesome parts (tracking must-usability through blocks) seem useful, but I could see them being added later on. I see the current solution as a 90% coverage, this RFC brings it to 99%, and the extras going to 99.9%.

> .ok().unwrap() makes me wonder what is wrong with their error type that results in .unwrap() not being used.

I have needed to do this when the Error type of the Result does not implement `Debug`. I find it annoying.

**Comment by huonw** [comment]
> You did not bother to show us the contents of your .cargo folders when you presented your quick analysis.

A good-faith guess at the contents of a cargo cache is likely to be much closer to what's there than a guess at the contents of some "random" folder called `rust`.

> This seems pretty clear.

Your line of thinking is clear, now that you have given some details.

In any case, I disagree. `Result` is designed for error handling, and getting help to avoid ignoring errors is great. It certainly seems that removing `#[must_use]` from `Result` would result in significantly more boilerplate in a significantly more error-prone way (remembering to add `#[must_use]` to almost all functions returning a `Result`) than needing to occasionally write `let _ =` or the bit of extra (reusable) code required to, e.g., define a `try!` equivalent for `Option` (or any other type).

Certainly, non-error-handling uses are a semantic abuse of `Result`, just like ignoring errors is one of `ok`.

**Comment by eddyb** [comment]
@shepmaster a type not implementing `Debug` is likely a bug.

**Comment by shepmaster** [comment]
> Certainly, non-error-handling uses are a semantic abuse of Result, just like ignoring errors is a semantic abuse of ok.

Does that indicate [`Entry::get`](http://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.get) should be changed to not return a `Result`? I'd be in favor of that!

**Comment by Gankro** [comment]
Calling `Entry::get` and not using the result *would* be a programming error. It has no side-effects. It is analogous to the `ok` method on Result itself-- a coercion to a different form. That said, I'm open to refactoring that API in the future.

**Comment by reem** [comment]
We could also have a `#[must_use]` and a `#[should_use]` where `must_use` generates an error and `should_use` generates a warning. You can use `must_use` for things where it is dangerous to proceed without investigating the error, and `should_use` for stuff like `Option`, which is often used for non-side-effectful operations, since they have few sources of failure.

**Comment by jakub-** [comment]
I'm in favour but it seems like the attribute should read `must_use_result` as that's what it really means. That'd make the name distinct from the type-level attribute that exists now but I think that's fine.

**Comment by jakub-** [comment]
With regards to having the `#[must_use]` annotation on `Result`, I think in the short-term it's preferable as  relaxing it to individual functions is a backwards-compatible change and the opposite is not.

I imagine that what we really want is having `#[must_use]` propagate structurally bottom-up through types in which case ideally `#[muse_use]` would apply to types implementing the `Error` trait. However, this is a rather subtle improvement given that I agree `Result` is meant to be used for error handling and not as a generic `Either`-like type.

**Comment by Thiez** [comment]
I think this talk of making `must_use` an error is silly. It is a helpful lint that will remind you to inspect some value if you have forgotten to do so, but it is easily defeated. Take this example:

```rust
(Ok::<bool,bool>(true));
(Ok::<bool,bool>(true),);
(Ok::<bool,bool>(true),Ok::<bool,bool>(true));
```

The first line gives the usual warning. The second line triggers no warning, because now we "use" this result by adding a comma to the end. Okay, we used it by putting it into a one-element tuple that we then discard. Same with the third line. Or we can just `drop` the damn thing. Really, the lint is there as a reminder, and turning those warnings into errors is pointless: if I *really* want to ignore a `Result` (or any other `[must_use]` type) I will find a way to do so. If I don't want to ignore a `Result` then the warning will be sufficient to convince me to change the code. Either way turning the warning into an error will only be inconvenient.

Rust already has many rules, but they tend to ensure the very useful property of memory safety, which makes having them an acceptable tradeoff. Ignoring results is not `unsafe` in the Rust meaning of the word. Let's not slide any further into "bondage and discipline"-language territory than we already are without a very good reason.

**Comment by mahkoh** [comment]
>Certainly, non-error-handling uses are a semantic abuse of Result

Like

http://doc.rust-lang.org/std/slice/trait.SliceExt.html#tymethod.binary_search_by,
http://doc.rust-lang.org/std/collections/btree_map/enum.Entry.html#method.get

@Gankro: Stop abusing Result with your APIs.

**Comment by huonw** [comment]
Could you leave off the snark? Just pointing out the APIs you think violate the statement would've been enough, no need to abuse @Gankro for it. (Anyway, see https://github.com/rust-lang/rfcs/pull/921.)

**Comment by mahkoh** [comment]
I see. That leaves the abuse of Result in `binary_search_by`. Are we going to introduce a new type for that function?

Mind you that I only did a quick check of libcollections.

@Gankro I'm sorry for any discomfort my abuse might have caused you.

File: text/0000-must-use-functions.md (hunk 5 above)
Comment (Manishearth) [inline text/0000-must-use-functions.md]: @oli-obk it doesn't look like this applies to types.

**Comment by Manishearth** [comment]
@Thiez 

> I think this talk of making must_use an error is silly.

Could you link to it? I don't see any such talk in this thread, it's a warning and will remain one (right? I may have missed something) unless a user explicitly turns it into an error.

**Comment by Thiez** [comment]
@Manishearth reem mentions it slightly above my comment.

File: text/0000-must-use-functions.md (hunk 4 above)
Comment (reem) [inline text/0000-must-use-functions.md]: We could add an ignore method to all types (with a blanket impl) that's just a method form of drop.

**Comment by Manishearth** [comment]
Ah, okay. I disagree with having error-producing lints, at least right now.

**Comment by oli-obk** [comment]
discussion has been at a stand-still for 2 months... I guess everyone said their piece and it should get accepted?

On that note: my comment somehow got under the wheels of the rest of the discussion:
> @huonw: could you address must_use traits (or at least forbid them explicitly for now)? Currently that attribute on traits does nothing (see http://is.gd/JMciVr)

**Comment by oli-obk** [comment]
ping @huonw

**Comment by aturon** [comment]
This RFC is one of those cases where the feature makes perfect sense (and is well-motivated) *in principle*, but I am unconvinced that it's worthwhile *in practice*. For example, I'm not 100% sure I'd want to label the `ok` method with `must_use`, even if we had the proposed feature. But, on the other hand, I feel like I've run across functions from time to time that return a "normal" type but would benefit from this `must_use` annotation.

Basically, I'm not opposed to this RFC, but I'm not in a hurry to accept it until we've accumulated a more substantial list of functions that would benefit from it.

**Comment by nikomatsakis** [comment]
Hear ye, hear ye. This RFC is now entering **final comment period**.

**Comment by Thiez** [comment]
I agree with @aturon that it would be nice to see a list of candidate functions/methods where the new attribute would be used. It seems a lot of attention has gone to how `Result::ok` is used in practice, but is there consensus that that method needs to be marked `#[must_use]`? Even if we assume there is, are there any additional places in the standard library that would benefit from this new attribute, or will this new feature be introduced to be used only on a single line?

**Comment by rkjnsn** [comment]
:+1: , and I would like to see cases like `(foo());` and `{ ...; foo() };` to be handled as well, if at all possible.

**Comment by nikomatsakis** [comment]
My feeling is that I still this RFC makes sense and I could see it being useful *but* I do not recall ever having personally felt the need for it.

**Comment by andrew-d** [comment]
:+1: on this - I can see this also being useful for, e.g. `collection.get(key)` vs `collection.exists(key)`, or something similar.

**Comment by Thiez** [comment]
@andrew-d Could you elaborate? I don't think the standard library contains any type that has both an `exists` and a `get` function, and even if I imagine you meant `HashMap`'s `get` and `contains_key` I can't think of a situation where a `#[must_use]` attribute would actually make a difference.

**Comment by andrew-d** [comment]
> @andrew-d Could you elaborate? I don't think the standard library contains any type that has both an `exists` and a `get` function, and even if I imagine you meant HashMap's `get` and `contains_key` I can't think of a situation where a `#[must_use]` attribute would actually make a difference.

Mostly just idle speculation, but I was thinking that for anything that had both a `get` and `exists`-style function, you could tag the `get` function as `#[must_use]`, to prevent someone from using the `get` form when the other would do.  Not thinking of any collection (standard library or otherwise) in particular.

**Comment by brson** [comment]
I'll echo @Thiez's sentiments that we should have some immediate use cases in std before going forward with this. It doesn't seem like the matter of whether `ok` will get this attribute is resolved.

**Comment by alexcrichton** [comment]
I, too, agree that it does not seem pressing to accept this RFC and I'd be fine not having it for awhile in std.

**Comment by nikomatsakis** [comment]
Consensus at the language subteam meeting was to close this RFC for now pending stronger motivation.

**Comment by fenhl** [comment]
See #1812 for some motivation.

**Comment by Philipp91** [comment]
More motivation: In an FFI, you get lots of `c_int`s as return values. As far as I can tell, you're in no position to change that (make it a `Result<c_int, ()>` or something like that) and you cannot flag `c_int` as `#[must_use]` in general. Still some of those return values are safe to ignore whereas others aren't. And in some cases (particularly when the FFI is generated from C code that contains some `__attribute__ ((warn_unused_result))` attributes) the information is easily available. Allowing `#[must_use]` would make this information usable and useful.

**Comment by joshtriplett** [comment]
This came up again in the context of #1812, which led to a proposal to close #1812 and re-open this.  Anyone interested in working on a revived RFC?

I would suggest focusing first on support for `#[must_use]` on functions, without specifying any particular functions to mark with that attribute.  We can then handle adding that attribute to specific functions afterward, likely with associated crater runs to determine the impact of doing so.

**Comment by withoutboats** [comment]
We have found a motivation for this in #1812 - `PartialEq::partial_eq`. We'd have to check the impact, but it seems clear that `foo == bar;` is the result of either some confusion on the author's part or a mistyped assignment.

**Comment by ssokolow** [comment]
@withoutboats Wow... if that doesn't have prohibitive impacts, it'd feel like a direct act of one-upmanship to Python's decision to sacrifice `if let`'s ancestors (eg. `if (entry = thing.pull()) {`) on the altar of typo-prevention... and that's a good thing.

(And if it *is* blocked by the stability promise, I think it's good reason to start explicitly maintaining a list of "we wish we'd thought of this sooner" features that people writing "Rust-inspired languages" (ie. Rust 2.0 without the Perl6/Py3k/etc. drama) should consider.)

Also, I think the point made by @Philipp91 deserves more attention.

Even without automatic translation of `__attribute__ ((warn_unused_result))`,  `#[must_use]` on functions feels like a great way to counterbalance the effects of C's primitive type system on `-sys` crate internals.

*With* automatic translation, it would help binding generators to have that "go above and beyond" feel that already draws praise in areas like `rustc`'s error messages.

**Comment by withoutboats** [comment]
@ssokolow since a) its a warn and not an error, b) you can fix it with `let _ = foo == bar;`, c) why would anyone ever write `foo == bar;`?, I'm optimistic that the impact will be small and acceptable.

We'd need an implementation and a crater run. I agree with @joshtriplett that we should accept & implement this without specifying which functions to apply it to, and then just do crater runs on applying it things (IMO without RFCs but YMMV).

**Comment by jmacdonald** [comment]
I'd love to see this, especially as a means to make a new `ignore` method on `Result` the idiomatic way to suppress these warnings. The syntax is much more intention revealing than `ok()` and `let _ = ...;`, and feels like a great fit, _especially for Rust applications_, where typical error propagation practices may not be the best fit (and the very existence of these other idioms seems to hint at a need for this sort of syntax).

**Comment by Thiez** [comment]
So you would *especially* like to have this feature so that a new method could be introduced specifically to avoid it? That doesn't sound very convincing to me 😛 

What exactly would make `foo().ignore();` more intention-revealing that `let _ = foo();`? At first sight it looks more obvious, but `let _ = ...` means "ignore" *everywhere*, while `.ignore()` would mean ignore *only* on `Result`. So it seems to me that the `let _ = ...` form is actually more intention revealing, because it means the same thing in all possible contexts.

**Comment by jmacdonald** [comment]
No, the benefit of this feature is _discouraging the use of `ok()/err()` to suppress must-use warnings_ (or warning about it as an unintended side-effect of their usage). 😄 

As for the `let _ = ...` idiom, I'd argue that it's just the _underscore_ that means ignore, and that its use is best suited to individual positional arguments, tuple element destructuring, or match branches that you'd want to ignore (where a placeholder _must_ exist). There's already an idiom for discarding a return value from an expression: the semi-colon! Introducing `let` to bind the result to a discarded variable feels odd, at least to me. If that's the idiom we want to encourage, it'd be great to add a reference to it in the [`Error Handling` doc section](https://doc.rust-lang.org/book/error-handling.html), or in the [upcoming book](http://rust-lang.github.io/book/ch09-00-error-handling.html).

Great point about this only applying to `Result`, though; you'd need something that applies to any type or function that could raise this warning.

**Comment by withoutboats** [comment]
I'm intrigued by the idea of `ignore()` instead of `let _ = `, because I think its much clearer, but it seems like its an idea to consider separately from this RFC.

This RFC needs some edits to get it into the current idea behind this proposal:

* Remove the references to `Result::ok` and `Result::err` (except maybe as a possible motivation)
* Add the `partial_eq` example to the motivation.

However, @huonw is no longer a contributor, even if someone else makes the edits I don't think we can get them onto this branch without his merging them. I don't know what the procedure here is, we may need to open a new PR after all.

If anyone wants to volunteer to make the edits (possibly @crumblingstatue ?) we could probably get this into FCP soon; based on the discussion here and on #1812, this seems pretty well-discussed & without outstanding issues.

**Comment by Wyverald** [comment]
> There's already an idiom for discarding a return value from an expression: the semi-colon!

Is that exactly the source of mistakes, though? (because the semi-colon is easily used to discard return values that we normally wouldn't want to discard)

A more radical way to look at this, IMO, would be to enforce that the expression preceding `;` *has* to have type `()`. This essentially means that nothing can be implicitly discarded by chaining `;`.

**Comment by arielb1** [comment]
> A more radical way to look at this, IMO, would be to enforce that the expression preceding `;` has to have type `()`. This essentially means that nothing can be implicitly discarded by chaining `;`.

There are several functions that return an "optional" result (IIRC some versions of `Vec::push` used to return a reference to the new element), and this would make using them hard.

**Comment by Wyverald** [comment]
Indeed, the "no implicit discard" rule wouldn't work very well with this kind of APIs, but maybe *those* APIs are the exception instead of the norm, and we should be annotating *them* with `#[can_ignore]` or something.

Just to clarify, I'm not pushing for the adoption of this rule -- it would be such a breaking change, after all. But I still think we might be overestimating how often return values need to be discarded implicitly, and if we were to design everything differently, maybe this rule could've been considered.

**Comment by iopq** [comment]
I also came here from #1812 - it's great motivation. I just saw a bug in Android source code in the form of `modem_reset_flag == 0;`

I realized Rust does NOT do better in this case

**Comment by joshtriplett** [comment]
I'd like to avoid derailing this RFC with even stricter proposals.  I think this RFC makes sense as a concrete step that would have minimal impact or invasiveness for existing code and projects.

Please, by all means, go ahead and propose a separate RFC (or perhaps a pre-RFC first) on a more universal "never ignore return values" mechanism; at a minimum, that might make sense as an off-by-default lint, even if not an on-by-default one.

**Comment by withoutboats** [comment]
As I said before, I'd like to fcp this rfc as soon as it's been amended to reflect the current status. If anyone can write up an amendment it would be great.

**Comment by nikomatsakis** [comment]
Historical note:

> A more radical way to look at this, IMO, would be to enforce that the expression preceding ; has to have type (). This essentially means that nothing can be implicitly discarded by chaining ;.

Many of us have been in favor of this in the past, but (imo) we definitively decided against it when we allowed APIs like `insert()` or `push()`. The notion of `#[must_use]` types was intended as the compromise. I do not think we should relitigate that decision but rather stay within the spirit (as this RFC indeed proposes to do) of generally permitting unused values but allowing types/functions to declare when we should warn.

**Comment by withoutboats** [comment]
Is anyone interested in updating this RFC to reflect the learnings from #1812 ? @crumblingstatue ? @joshtriplett ?

**Comment by withoutboats** [comment]
Just to make things simpler I'm closing this and making #1940 the PR for tracking this RFC.
